src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.cs

```diff
@@ -2358,36 +2345,20 @@ internal override int MetadataArity
                 }
             }
 
-            private readonly PENamedTypeSymbolNonGeneric _underlying;
-            internal override bool IsNativeInt { get; }
-
             internal override NamedTypeSymbol AsNativeInt(bool asNativeInt)
             {
                 Debug.Assert(this.SpecialType == SpecialType.System_IntPtr || this.SpecialType == SpecialType.System_UIntPtr);
 
-                if (IsNativeInt == asNativeInt)
-                {
-                    return this;
-                }
-
-                return asNativeInt ? new PENamedTypeSymbolNonGeneric(this, isNativeInt: true) : _underlying;
+                return asNativeInt ?
+                    (NamedTypeSymbol)new NativeIntegerTypeSymbol(this) : // PROTOTYPE: Consider caching instance, perhaps on containing assembly.
+                    this;
             }
 
-            // PROTOTYPE: Temporary approach for AsNativeInt().
             internal override bool Equals(TypeSymbol t2, TypeCompareKind comparison, IReadOnlyDictionary<TypeParameterSymbol, bool> isValueTypeOverrideOpt = null)
             {
-                if ((object)t2 == null) return false;
-                if ((object)t2 == this) return true;
-
-                var other = t2 as PENamedTypeSymbolNonGeneric;
-                if ((object)other != null &&
-                    (this.IsNativeInt || other.IsNativeInt) &&
-                    this.SpecialType == other.SpecialType)
-                {
-                    return true;
-                }
-
-                return base.Equals(t2, comparison, isValueTypeOverrideOpt);
+                return t2 is NativeIntegerTypeSymbol nativeInteger ?
```

Comment:
This `Equals` bothers me a little, but I'm not sure why. The dispatch on type feels like it would cause problems as we add more wrapper types.
Do we really need to be able to compare an underlying type (`PENamedTypeSymbol`) with a wrapped type (`NativeIntegerTypeSymbol`)?

Reply:
> Do we really need to be able to compare an underlying type (PENamedTypeSymbol) with a wrapped type (NativeIntegerTypeSymbol)?

Yes, in many cases, `System.IntPtr` should be considered equivalent to `nint`, similar to the behavior for `object` and `dynamic`.

---
In reply to: [374992403](https://github.com/dotnet/roslyn/pull/40994#discussion_r374992403) [](ancestors = 374992403)